system_hooks/src/l2_base_token.rs

```diff
@@ -140,206 +155,317 @@ where
-                return Ok(Err(
-                    "L2 base token failure: withdraw called with invalid calldata",
-                ));
-            }
-            l1_messenger_calldata[68..88].copy_from_slice(&calldata[(4 + 12)..36]);
-            l1_messenger_calldata[88..120]
-                .copy_from_slice(&nominal_token_value.to_be_bytes::<32>());
-
-            let result = send_to_l1_inner(
-                &l1_messenger_calldata,
-                resources,
-                system,
-                L2_BASE_TOKEN_ADDRESS,
-                caller_ee,
-            )?;
-
-            Ok(result.map(|_| &[] as &[u8]))
+        s if s == WITHDRAW_SELECTOR => withdraw(
+            calldata,
+            calldata_len,
+            resources,
+            system,
+            caller,
+            caller_ee,
+            nominal_token_value,
+            is_static,
+        ),
+        s if s == WITHDRAW_WITH_MESSAGE_SELECTOR => withdraw_with_message(
+            calldata,
+            calldata_len,
+            resources,
+            system,
+            caller,
+            caller_ee,
+            nominal_token_value,
+            is_static,
+        ),
+        _ => Ok(Err("L2 base token: unknown selector")),
+    }
+}
+
+fn withdraw<S: EthereumLikeTypes>(
+    calldata: &[u8],
+    calldata_len: u32,
+    resources: &mut S::Resources,
+    system: &mut System<S>,
+    caller: B160,
+    caller_ee: u8,
+    nominal_token_value: U256,
+    is_static: bool,
+) -> Result<Result<&'static [u8], &'static str>, SystemError>
+where
+    S::IO: IOSubsystemExt,
+{
+    if is_static {
+        return Ok(Err(
+            "L2 base token failure: withdraw called with static context",
+        ));
+    }
+    // following solidity abi for withdraw(address)
+    if calldata_len < 36 {
+        return Ok(Err(
+            "L2 base token failure: withdraw called with invalid calldata",
+        ));
+    }
+
+    burn_nominal_token_value(resources, system, caller_ee, &nominal_token_value)?;
+
+    // Sending L2->L1 message.
+    // ABI-encoded messages should consist of the following:
+    // 32 bytes offset (must be 32)
+    // 32 bytes length of the message
+    // followed by the message itself, padded to be a multiple of 32 bytes.
+    // In this case, it is known that the message is 56 bytes long:
+    // - IMailbox.finalizeEthWithdrawal.selector (4)
+    // - l1_receiver (20)
+    // - nominal_token_value (32)
+
+    // So the padded message will be 64 bytes long.
+    // Total length of the encoded message will be 32 + 32 + 64 = 128 bytes.
+    let mut l1_messenger_calldata = [0u8; 128];
+    l1_messenger_calldata[31] = 32; // offset
+    l1_messenger_calldata[63] = 56; // length
+    l1_messenger_calldata[64..68].copy_from_slice(FINALIZE_ETH_WITHDRAWAL_SELECTOR);
+    // check that first 12 bytes in address encoding are zero
+    // TODO: should we fail or silently truncate? Solidity would truncate iirc.
```

Comment:
@AntonD3 probably we should change it. Wdyt?

Reply:
I think it makes sense to follow solidity abi decoder behavior here, worth to check